nx/lib/nx.ex

```diff
@@ -1540,22 +1543,29 @@ defmodule Nx do
           ]
         ]
       >
+  
+  You can use `:auto` to infer dimension sizes. This is useful when you
+  don't know the size some dimension should be ahead of time:
 
+      iex> t= Nx.tensor([[[1, 2, 3]], [4, 5, 6]])
+      iex> Nx.reshape({:auto, 2}, names: [:x, :y])
+      #Nx.Tensor<
+        s64[x: 3][y: 2]
+        [
+          [1, 2],
+          [3, 4],
+          [5, 6]
+        ]
+      >
   """
   @doc type: :shape
   def reshape(tensor, new_shape, opts \\ []) do
     %T{shape: old_shape} = tensor = to_tensor(tensor)
     new_names = opts[:names] || names!(new_shape)
-    new_shape = shape(new_shape)
```

Comment:
We allow the second argument to be a tensor that we extract the shape from. So I think we need to keep this line (and perhaps add a test/doctest for this behaviour). This should probably be something like this:

    new_shape = if is_tuple(new_shape), do: new_shape, else: shape(new_shape)